src/main/java/com/rabbitmq/http/client/ReactorNettyClient.java

```diff
@@ -511,98 +503,101 @@ public static String basicAuthentication(String username, String password) {
     }
 
     private <T> Mono<T> doGetMono(Class<T> type, String... pathSegments) {
-        return client.get(uri(pathSegments), request -> Mono.just(request)
-            .transform(this::addAuthorization)
-            .flatMap(HttpClientRequest::send))
-            .onErrorMap(this.errorHandler)
-            .transform(decode(type));
+        return client.headers(authorizedHeader())
+            .doOnResponse(onResponseCallback)
+            .get()
+            .uri(uri(pathSegments))
+            .responseSingle(decode(type));
+    }
+
+    protected <T> BiFunction<? super HttpClientResponse, ? super ByteBufMono, ? extends Mono<T>> decode(Class<T> type) {
+        return (response, byteBufFlux) -> {
+            if (response.status().code() == 404) {
+                return Mono.empty();
+            } else {
+                return byteBufFlux.asByteArray().map(bytes -> {
+                    try {
+                        return objectMapper.readValue(bytes, type);
+                    } catch (IOException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                });
+            }
+        };
     }
 
     @SuppressWarnings("unchecked")
     private <T> Flux<T> doGetFlux(Class<T> type, String... pathSegments) {
         return (Flux<T>) doGetMono(Array.newInstance(type, 0).getClass(), pathSegments).flatMapMany(items -> Flux.fromArray((Object[]) items));
     }
 
+    protected Consumer<HttpHeaders> authorizedHeader() {
+        return headers -> {
+            // FIXME try to get the token as late as possible?
```

Comment:
the block call here is a problem. We have been requested to add something like headersWhen(Function<HttpHeaders, Mono<HttpHeaders>>) to resume the request when the headers are resolved.

Reply:
@smaldini FYI I use now `HttpClient#headersWhen` to get the authentication token as late as possible.